Added support for Argo WREM-3 A/C remote protocol [part1]

mbronk opened the pull request:
Adds support for ARGO WREM-3 remote

For: https://github.com/crankyoldgit/IRremoteESP8266/issues/1912

This PR is a standalone _part1_ of changes detached from: https://github.com/crankyoldgit/IRremoteESP8266/pull/1913

mbronk committed:
Added support for Argo WREM-3 A/C remote protocol

Signed-off-by: Mateusz Bronk <bronk.m+gh@gmail.com>

mbronk committed:
Added additional UT for Argo AC via IRac/common

Incl. a macro 'gadget' to allow capturing raw command sent by IRac.

Signed-off-by: Mateusz Bronk <bronk.m+gh@gmail.com>

crankyoldgit commented on Doxyfile:
```diff
@@ -10,7 +10,8 @@ EXTRACT_ALL            = YES
 EXTRACT_LOCAL_CLASSES  = NO
 GENERATE_LATEX         = NO
-ENABLE_PREPROCESSING   = NO
+ENABLE_PREPROCESSING   = YES
```
Why do you need to do this change? This will affect the documentation for other board types.

mbronk replied:
The trigger for making this change was a build failure, actually. 
Without it, documentation for `OUTPUT_DECODE_RESULTS_FOR_UT` 
https://github.com/mbronk/IRremoteESP8266/blob/4707a4d0a236f6c71953626f87d2998640c4d4ad/src/IRac.cpp#L70 yields
```
/IRac.cpp:93: error: found documented #define but ignoring it because ENABLE_PREPROCESSING is NO.
 (warning treated as error, aborting now)
```

---
Now, this is a poor reason to change it, of course... though I figured it is a good change on its own (hence detached to separate commit), as it makes Doxygen `#ifdef` aware (which is also why I had to change in a few other files, as it begun to complain of functions being wrongly documented (rightfully so).
I have these changes decoupled to a separate commit (https://github.com/mbronk/IRremoteESP8266/commit/03767d800c019aff15cc26e2a39e3282c4fd5525) so may detach to a separate PR or skip completely (though on the latter, I'd argue it might be a good direction to enable it eventually).
Let me know which one you prefer.

crankyoldgit replied:
Honestly, I'd love it to handle preprocessor stuff better/correctly, but it doesn't for most of the library.
Having the documentation favour one processor vs. the  other e.g. ESP8266  vs ESP32 isn't really good IMHO.
I'd rather have this bit of code skipped doxygen-wise than break some of the other documentation assumptions.

If you have a better solution, I'm all ears.

mbronk replied:
> Having the documentation favour one processor vs. the other e.g. ESP8266 vs ESP32 isn't really good IMHO.

Agreed. On the flip side, I guess it's already happening somewhat anyway. Ex.:
https://crankyoldgit.github.io/IRremoteESP8266/doxygen/html/classIRrecv.html#a8fe4d26ef1f863db1db9994fed5fc209
documents https://github.com/crankyoldgit/IRremoteESP8266/blob/55ce00262bcace187c97be14aca6536fa53bbd99/src/IRrecv.cpp#L267-L269

...which is ESP32-specific (though the documentation is deliberately written this way).

Since it's a bigger topic, in order to move this one forward, I'll **revert the Doxy changes for now**, and possibly reintroduce later as its own PR if I can think of a way of making it more proper.

BTW. the `ESP32=1` is not really doing anything useful, it's the `ENABLE_PREPROCESSING=YES` which makes it more "_natural_" (less ```///@cond``` etc.), but I agree this likely requires more caution to introduce (merely flipping it to `YES` makes a few errors appear due to how certain functions are documented (commit 03767d800c019aff15cc26e2a39e3282c4fd5525 shows these).

crankyoldgit replied:
As discussed. Let move that discussion/change to a different PR. No need to hold this up.

crankyoldgit commented on src/IRutils.cpp:
```diff
@@ -722,10 +729,12 @@ namespace irutils {
-    String result = addIntToString(degrees, kTempStr, precomma);
+                         const bool precomma, const bool isRoomTemp) {
+    String result = addIntToString(degrees, (isRoomTemp)? kRoomStr : kTempStr,
+                                   precomma);
```
I think elsewhere we use 'Sensor' for the "room temperature". i.e. That is what the remote is reporting, it's sensor.
Please use `kSensorStr` etc for this for consistancy.

crankyoldgit commented on src/IRutils.cpp:
```diff
@@ -736,12 +745,14 @@ namespace irutils {
   /// @return The resulting String.
   String addTempFloatToString(const float degrees, const bool celsius,
-                              const bool precomma) {
+                              const bool precomma, const bool isRoomTemp) {
```
See other comment.

crankyoldgit commented on src/IRutils.cpp:
```diff
@@ -736,12 +745,14 @@ namespace irutils {
     result.reserve(14);  // Assuming ", Temp: XXX.5F" is the largest.
-    result += addIntToString(degrees, kTempStr, precomma);
+    result += addIntToString(degrees, isRoomTemp? kRoomStr : kTempStr,
+                             precomma);
```
Sensor

mbronk replied:
This should be resolved now

crankyoldgit commented on src/IRutils.h:
```diff
@@ -60,16 +60,19 @@ namespace irutils {
                         const uint8_t quiet, const uint8_t medium,
-                        const uint8_t maximum = 0xFF);
+                        const uint8_t maximum = 0xFF,
+                        const uint8_t medium_high = 0xFE);
```
`0xFF` should be fine. It basically stands for not used. As something earlier has the same value, it won't be displayed/used etc.

crankyoldgit commented on src/ir_Argo.cpp:
```diff
@@ -468,8 +1353,364 @@ String IRArgoAC::toString(void) const {
+  return argo_ac_remote_model_t::SAC_WREM3;
+}
+
+namespace {
```
You might want to consider moving these to some place more central. i.e. IRutils or IRac etc.
They might come in handy for other protocols.

crankyoldgit commented on src/ir_Argo.cpp:
```diff
@@ -505,4 +1748,98 @@ bool IRrecv::decodeArgo(decode_results *results, uint16_t offset,
+                  kArgoBitMark, kArgoOneSpace,
+                  kArgoBitMark, kArgoZeroSpace,
+                  kArgoBitMark, kArgoGap,  // difference vs decodeArgo
+                  true, _tolerance, kArgoOverhead,
```
Are you sure you understand what `kArgoOverHead` (`excess`) does here? This seems quite odd.
See: https://crankyoldgit.github.io/IRremoteESP8266/doxygen/html/classIRrecv.html#a4448c1658383962d735353352987c9aa

mbronk replied:
Thanks for pointing that out! This now seems to me to be a bad refactor or something...And as far as the "am I sure" part... nah, not at all. I only know what I inferred about it by reading the code, so might be missing the point entirely.

----

I remember fiddling with all these parameters at some point, as I had a hit-and-miss real-life decode results, but for the life of me, I can't recall why I have changed this particular value. 

I'll revert it to `kMarkExcess` for now (UTs are passing, and these are real captures, so it can't be that bad :) ). 
If I later recall why I have changed it, or I see issues with it in practice, will raise as a separate PR.

crankyoldgit replied:
As it's working with `3`, you can just set (try) the excess to `0`.

mbronk replied:
Done

crankyoldgit commented on src/ir_Argo.cpp:
```diff
@@ -505,4 +1748,98 @@ bool IRrecv::decodeArgo(decode_results *results, uint16_t offset,
+/// @return True if the message is a WREM-3 one
+bool IRArgoAC_WREM3::isValidWrem3Message(const uint8_t state[],
+                                         const uint16_t nbits, bool strict) {
+  auto stateLength = std::min(static_cast<uint16_t>(std::ceil(nbits / 8.0)),
```
No `auto` please. use `uint16_t`.
See: https://github.com/crankyoldgit/IRremoteESP8266/wiki/Library-Maintainers-Guide#c

crankyoldgit commented on src/ir_Argo.cpp:
```diff
@@ -505,4 +1748,98 @@ bool IRrecv::decodeArgo(decode_results *results, uint16_t offset,
+/// @return True if the message is a WREM-3 one
+bool IRArgoAC_WREM3::isValidWrem3Message(const uint8_t state[],
+                                         const uint16_t nbits, bool strict) {
+  auto stateLength = std::min(static_cast<uint16_t>(std::ceil(nbits / 8.0)),
```
Why use floatingpoint math here?
You could check if it's a multiple of 8 (`% 8`) , if it isn't fail, and then just do a `div`/`/`.

mbronk replied:
The point here was, I really wanted to count byte-wise with excess (round to next full byte) and operate on raw bits a bit longer (when initially written, I wasn't 100% sure if I won't need to go lower than bytes - this code would technically be able to swallow a 6-bit message etc.).

Anyways - it makes no practical sense now, so will change as you suggested

crankyoldgit replied:
Thanks.

crankyoldgit commented on src/ir_Argo.cpp:
```diff
@@ -505,4 +1748,98 @@ bool IRrecv::decodeArgo(decode_results *results, uint16_t offset,
+    return false;
+  }
+
+  auto messageType = IRArgoACBase<ArgoProtocolWREM3>::getMessageType(state,
```
See `auto` comment earlier. This applies to everything in the core library. Unit test code excepted.

mbronk replied:
Got it, will make explicit.
Not sure if I get the comment though - i.e. `messageType` is of an enum type, so it hardly violates `All internal library code must use [c99 exact-width type definitions](https://en.wikipedia.org/wiki/C_data_types#Fixed-width_integer_types). e.g. uint16_t instead of unsigned int etc.` 

----
Just in case, I'll remove all uses of `auto`, but let me add the following (opinioated!) thoughts :) :
- If you'd like to avoid `auto` entirely in the core library, I strongly suggest adding a new note to the code guidelines, banning its use **explicitly**.
- It's a heavily opinioated subject, but FYI, I'm in the "_use of auto is often **superior** to writing explicit types as it allows the compiler to do its magic_" camp. See also: https://herbsutter.com/2013/06/13/gotw-93-solution-auto-variables-part-2/. Aka. I am removing it, but with a heavy heart :)

crankyoldgit replied:
Re: Adding note. I will. Thanks for the suggestion.

Oh I get using `auto` is much better, but to better simulate what's happening on an ESP platform, I need ensure the bit-lengths of integers. We've had this bite more than once. i.e. `int` being 16 bits on the platform and 32 on x86 land etc.
`auto`s should be fine for a simple iterator that isn't used in any of the loop's body for bit arithmatic. It's just safest to not have it so the unit tests pick up problems vs. compiler optimisations. I wish I could have both worlds. :-(

mbronk replied:
Just for the record - I fully get and embrace the need for having fixed-size integers on any ABI or HW interface. 
The comment you added though was related to a fully qualified fixed size enum (referred to via auto), which is what I reacted to :)
 In any case, it's gone now, so I think we can safely close this side-topic :)

crankyoldgit commented on src/ir_Argo.cpp:
```diff
@@ -505,4 +1748,98 @@ bool IRrecv::decodeArgo(decode_results *results, uint16_t offset,
+      break;
+    default:
+      return false;
+  }
```
Divide by 8 much earlier to get the bytes. And do the comparisons via that. That will reduce the resultant code size, and improve readablity. e.g. `statelength`.

crankyoldgit replied:
Also, this really is a "strict" condition IMHO.

mbronk replied:
> That will reduce the resultant code size, and improve readablity. e.g. `statelength`.

Good point - will simplify (agree with the readability part of the comment).
Must add though, that I'd be very surprised it does anything positive to the resulting code size. 
Compiler should be perfectly capable of optimizing it on its own.

----

> Also, this really is a "strict" condition IMHO.

Well, yes and no.
That is, it is about the only thing separating `WREM2` from `WREM3` (and WREM3 detection must run before WREM2 non-strict, as it is a swallow-all).
While there's indeed a `getMessageType()` doing the detect, it only takes 2 **bits** of input, so a chance of a false-positive is way high. Aka. I consider "checksum" strict check, but anything else... I deliberately wanted to fall through to WREM2 and not get handled here.

That said, I only use the detect in strict mode anyway, so the point I'm making above is a bit moot and I could equally well remove 'strict' param entirely...
Leaving as-is for now, let me know where you'd like to take it.

crankyoldgit replied:
My unwritten policy for `strict` is, if it is in some way protocol-format related and can cause it to abort decoding, encapsulate it in `strict`. The purpose is to allow more loose matching with `strict` set to `false`, and/or code re-use where the protocol timings etc are compatible, but the length or structure of the data differs in a different protocol.
i.e. Maximum potential for code-reuse to reduce library binary size etc.

mbronk replied:
Please note though that `WREM2` and `WREM3` are variants (models) of the same `ARGO` IR protocol. 
The difference is in message decoding and parsing, so in the case you describe:
>... but the length or structure of the data differs in a different protocol....

...there's already a `decodeArgo(..., strict=false)` which would accomplish that result just fine.
The `WREM3` flavor was meant signify this is semantically a different sub-proto.

If I make the length-check non-strict, then, arguably, `decodeArgoWREM3(..., strict=false)` would be almost no different from `decodeArgo(..., strict=false)`... which makes it a bit redundant IMHO (but arguably consistent)

My preferences (from most-preferred):
1. Leave as is (possibly less semantically consistent, but the most versatile impl.)
2. Remove `strict` param from `decodeArgoWREM3` altogether
3. Make the length non-strict (it has a potential to clash with wrem2, but since I'm not calling it this way anyway... I guess the user would know what they're after). 
I.e. when you call it with `strict=false` a `true` response no longer means there's any confidence in it really being a WREM3.

Your pick? :)

crankyoldgit replied:
Users typically can't call the `decodeProtocol()` routines directly. So, it's just us "internal" users.
Practically every `decodeProtocol()` has a `strict`, so I'd like to keep that calling convention.
Of the options, **3.** sounds best to me. It is almost always a given if a `decodeProtocol()` is called with `strict=false` that it's not a confident match. It assumes the caller knows what they are doing.

crankyoldgit commented on src/ir_Argo.h:
```diff
@@ -164,16 +342,21 @@ class IRArgoAC {
   uint8_t getTemp(void) const;
 
+  void setRoomTemp(const uint8_t degrees);
+  uint8_t getRoomTemp(void) const;
```
Please use SensorTemp etc.

crankyoldgit commented on src/IRtext.h:
```diff
@@ -116,6 +120,7 @@ extern IRTEXT_CONST_PTR(kIndirectStr);
 extern IRTEXT_CONST_PTR(kInsideStr);
 extern IRTEXT_CONST_PTR(kIonStr);
 extern IRTEXT_CONST_PTR(kJkeStr);
+extern IRTEXT_CONST_PTR(kKeyStr);
```
We've used `kComandStr` for other remotes, rather than add another string. Please use that if possible/practical.

mbronk replied:
This was meant not for a "remote key" or a "command" but rather the setting value (used as `Key:Value` concept).
used as:
```cpp
  case argoIrMessageType_t::CONFIG_PARAM_SET:
    result += addIntToString(_.config.Key, kKeyStr);
    result += addIntToString(_.config.Value, kValueStr);
    break;
```

I can change it to `Setting` to be more descriptive, but this would make it less reusable, I think.
Hence, leaving as-is for now, please advise.

crankyoldgit replied:
Sorry, I missed that. I thought it was used for a button/key-press etc. Ignore my feedback.

crankyoldgit commented on src/locale/defaults.h:
```diff
@@ -378,6 +378,9 @@ D_STR_INDIRECT " " D_STR_MODE
 #endif  // D_STR_MEDIUM
+#ifndef D_STR_MED_HIGH
+#define D_STR_MED_HIGH "Med-high"
+#endif  // D_STR_MEDIUM
```
Please construct this as a compound of the `D_STR_MED` & `D_STR_HIGH` as it will make internationalisation supported easier/work better.

crankyoldgit commented on src/locale/defaults.h:
```diff
@@ -445,6 +448,27 @@ D_STR_INDIRECT " " D_STR_MODE
+#endif  // D_STR_AC_COMMAND
+#ifndef D_STR_COMMAND_TYPE
+#define D_STR_COMMAND_TYPE "A/C Command"
+#endif  // D_STR_COMMAND_TYPE
```
perhaps just "Config" and "Command"?
The "Command" string already exists (less flash space)
And the "A/C" is a bit redundant IMHO.

crankyoldgit reviewed (changes requested):
Again, thanks for the massive effort. I understand what it's like.

Thanks for splitting it up too.

mbronk committed:
Review changes

Signed-off-by: Mateusz Bronk <bronk.m+gh@gmail.com>

mbronk commented:
@crankyoldgit  Thanks so much for reviewing! I know it wasn't a pleasant read, given the volume of changes in Argo class.

I think I addressed most of your comments in the new update. On a few comments that I left behind, I added replies or questions to you. Please take a look. 
I'd also appreciate if you mark the conversations as resolved where you believe the changes address your comments, so that I know if there's anything withstanding. Where I did not comment, I merely incorporated the proposal as-is (hopefully didn't miss any).
For now I'm leaving it as-is and waiting for your inputs as to any follow-ups needed.
Thanks!

crankyoldgit reviewed (commented):
Looking a lot better. Just a couple of minor things left.

mbronk commented:
Reverted the doc/doxy changes. Unfortunately, I wanted to do it (too) cleanly and rebased the whole thing, not to show the add + revert... which seems to have thrown github's incremental diff off :(   Sorry for that!

Bottom line, as far as I can tell the only remaining comment from the original round would be the one about "strict" mode. Please take a look

crankyoldgit commented:
Let me know when the `strict` changes are in as discussed, and I'm fine to approve & merge.

mbronk commented:
> Let me know when the `strict` changes are in as discussed

Just pushed. Please take a look.
Note I have pushed the lax logic into `decodeArgoWREM3()` (if not strict, bypasses call to `isValidWrem3Message` entirely).
Reason being: if I were to not make the length check in `isValidWrem3Message`, it becomes a `return true`.

crankyoldgit commented:
> > Let me know when the `strict` changes are in as discussed
> 
> Just pushed. Please take a look. Note I have pushed the lax logic into `decodeArgoWREM3()` (if not strict, bypasses call to `isValidWrem3Message` entirely). Reason being: if I were to not make the length check in `isValidWrem3Message`, it becomes a `return true`.

Good. 'Cause that what I'd expect.

crankyoldgit reviewed (approved):
As discussed, LGTM
